devito/ir/ietxdsl/cluster_to_ssa.py

```diff
@@ -41,117 +58,70 @@ class ExtractDevitoStencilConversion:
-            eq.lhs.function._C_name,
-        )
-        self.block.add_op(stencil_op)
-        self.block = stencil_op.block
-
-        # dims -> ssa vals
-        perf("Apply time offsets")
-        time_offset_to_field: dict[str, SSAValue] = {
-            i: stencil_op.block.args[i] for i in range(time_size - 1)
-        }
-
-        # reset loaded values
-        self.loaded_values: dict[tuple[int, ...], SSAValue] = dict()
-
-        # add all loads into the stencil
-        perf("Add stencil Loads")
-        self._add_access_ops(retrieve_indexed(eq.rhs), time_offset_to_field)
-
-        # add math
-        perf("Visiting math equations")
-        rhs_result = self._visit_math_nodes(eq.rhs)
-
-        # emit return
-        offsets = _get_dim_offsets(eq.lhs, self.time_offs)
-
-        assert (
-            offsets[0] == time_size - 1
-        ), "result should be written to last time buffer"
-        assert all(
-            o == 0 for o in offsets[1:]
-        ), f"can only write to offset [0,0,0], given {offsets[1:]}"
-
-        self.block.add_op(stencil.ReturnOp.get([rhs_result]))
-        outermost_block.add_op(func.Return())
-
-        return func.FuncOp.from_region(
-            "apply_kernel", [], [], Region([outermost_block])
-        )
-
-    def _visit_math_nodes(self, node: Expr) -> SSAValue:
+        output_function = eq.lhs.function
+        grid: Grid = output_function.grid
+        step_dim = grid.stepping_dim
+
+        # Get all functions used in the equation
+        functions = OrderedSet(*(f.function for f in retrieve_function_carriers(eq)))
+        # We identify time buffers by their function and positive time offset.
+        # We store a list of those here to help the following steps.
+        self.time_buffers = [(f, i) for f in functions for i in range(f.time_size)]
+        # Also store the time buffer used for output in this equation
+        output_time_offset = (eq.lhs.indices[step_dim] - step_dim) % eq.lhs.function.time_size
+        self.out_time_buffer = (output_function, output_time_offset)
+
+
+        # For each used function, define as many fields as its time_size
+        fields_types = [field_from_function(f) for (f, _) in self.time_buffers]
+        # Create a function with the fields as arguments
+        xdsl_func = func.FuncOp("apply_kernel", (fields_types, []))
+        # Define nice argument names to try and stay sane while debugging
+        # And store in self.function_args a mapping from time_buffers to their
+        # corresponding function arguments.
+        self.function_args = {}
+        for i, (f, t) in enumerate(self.time_buffers):
+            xdsl_func.body.block.args[i].name_hint = f"{f.name}_vec_{t}"
+            self.function_args[(f,t)] = xdsl_func.body.block.args[i]
+
+        with ImplicitBuilder(xdsl_func.body.block):
+            self._build_step_loop(step_dim, eq)
+            # func wants a return
+            func.Return()
+
+        return xdsl_func
```

Comment:
empty line

Reply:
How was this one not requested by flake? :o